src/dataflow/components/stats.tsx

```diff
@@ -0,0 +1,276 @@
+  runProgramDataDocuments: number;
+  runProgramRelayOnlyDocuments: number;
+  publishedProgramDocuments: number;
+}
+
+interface IProps extends IBaseProps {}
+
+interface IState {
+  loading: boolean;
+  userStats: StatsGroup[];
+}
+
+const kHeaderNames = ["Documents", "Editable Programs", "Run Programs",
+                      "Run Programs w/ Data", "Run Programs Relay-Only", "Published Programs"];
+
+@inject("stores")
+@observer
+export class StatsComponent extends BaseComponent<IProps, IState> {
+
+  constructor(props: IProps) {
+    super(props);
+    this.state = {
+      loading: true,
+      userStats: [],
+    };
+  }
+
+  public async componentDidMount() {
+    const {db} = this.stores;
+    // request the published documents
+    const publicationsPath = db.firebase.getClassPersonalPublicationsPath(db.stores.user);
+    const publicationsRef = db.firebase.ref(publicationsPath);
+    const publicationsSnapshot = await publicationsRef.once("value");
+    const publishedDocOriginIds: string[] = [];
+    if (publicationsSnapshot) {
+      const pulishedDocs = publicationsSnapshot.val();
+      Object.keys(pulishedDocs).forEach((pulishedDoc) => {
+        const pulishedDocEntry = pulishedDocs[pulishedDoc];
+        if (!publishedDocOriginIds.find(originDoc => originDoc === pulishedDocEntry.originDoc)) {
+          publishedDocOriginIds.push(pulishedDocEntry.originDoc);
+        }
+      });
+    }
+
+    // request the users
+    const usersPath = db.firebase.getUsersPath(db.stores.user);
+    const usersRef = db.firebase.ref(usersPath);
+    const usersSnapshot = await usersRef.once("value");
+    if (usersSnapshot) {
+      const users = usersSnapshot.val();
+      const userStats: StatsGroup[] = [];
+
+      Object.keys(users).forEach((userId) => {
+        const user = this.stores.class.getUserById(userId);
+        if (user && user.type === "student") {
+          const statsGroup: StatsGroup = {
+            studentName: user.fullName,
+            documents: 0,
+            programDocuments: 0,
+            runProgramDocuments: 0,
+            runProgramDataDocuments: 0,
+            runProgramRelayOnlyDocuments: 0,
+            publishedProgramDocuments: 0
+          };
+
+          const userEntry = users[userId];
+          Object.keys(userEntry.personalDocs).forEach((pDoc) => {
+            const document = userEntry.personalDocs[pDoc];
+
+            // determine if this document is a published document
+            if (publishedDocOriginIds.find(originDoc => originDoc === document.self.documentKey)) {
+              statsGroup.publishedProgramDocuments++;
+            }
+
+            // determine if this is a run program
+            if (document.properties && document.properties.dfRunId) {
+              statsGroup.runProgramDocuments++;
+              document.properties.dfHasData
+                ? statsGroup.runProgramDataDocuments++
+                : statsGroup.runProgramRelayOnlyDocuments++;
```

Comment:
Not an engineering decision, but I think it would make more sense to report the number of programs with relays, as the number of relay-only documents can be derived, but the reverse is not true.

Reply:
Good call!  I added a column for the number of programs with relays and added it to the stat group.  
One note - with recent documents we can derive the number of relay-only programs.  However, we have a bunch of older documents floating around that do not have all of the current properties and can result in some odd results (especially if we try to derive the numbers from the other values).  For now, I'm going to insist that a relay-only program must have the `dfHasRelay` property and not have the `dfHasData` property).  Older documents that might be missing some of the properties will be ignored in the counts for now (rather than making an assumption).